source/client_channel_handler.c

```diff
@@ -531,103 +595,15 @@ static int s_process_read_message(
-                (void *)connection);
-
-            goto cleanup;
-        }
-
-        /* Handle the completed pending packet */
-        struct aws_byte_cursor packet_data = aws_byte_cursor_from_buf(&connection->thread_data.pending_packet);
-        AWS_LOGF_TRACE(AWS_LS_MQTT_CLIENT, "id=%p: full mqtt packet re-assembled, dispatching.", (void *)connection);
-        result = s_process_mqtt_packet(connection, aws_mqtt_get_packet_type(packet_data.ptr), packet_data);
-
-    handle_error:
-        /* Clean up the pending packet */
-        aws_byte_buf_clean_up(&connection->thread_data.pending_packet);
-        AWS_ZERO_STRUCT(connection->thread_data.pending_packet);
-
-        if (result) {
-            return AWS_OP_ERR;
-        }
-    }
-
-    while (message_cursor.len) {
-
-        /* Temp byte cursor so we can decode the header without advancing message_cursor. */
-        struct aws_byte_cursor header_decode = message_cursor;
-
-        struct aws_mqtt_fixed_header packet_header;
-        AWS_ZERO_STRUCT(packet_header);
-        int result = aws_mqtt_fixed_header_decode(&header_decode, &packet_header);
-
-        /* Calculate how much data was read. */
-        const size_t fixed_header_size = message_cursor.len - header_decode.len;
-
-        if (result) {
-            if (aws_last_error() == AWS_ERROR_SHORT_BUFFER) {
-                /* Message data too short, store data and come back later. */
-                AWS_LOGF_TRACE(
-                    AWS_LS_MQTT_CLIENT, "id=%p: message is incomplete, waiting on another read.", (void *)connection);
-                if (aws_byte_buf_init(
-                        &connection->thread_data.pending_packet,
-                        connection->allocator,
-                        fixed_header_size + packet_header.remaining_length)) {
-
-                    return AWS_OP_ERR;
-                }
+    int result = aws_mqtt311_decoder_on_bytes_received(&connection->thread_data.decoder, message_cursor);
 
-                /* Write the partial packet. */
-                if (!aws_byte_buf_write_from_whole_cursor(&connection->thread_data.pending_packet, message_cursor)) {
-                    aws_byte_buf_clean_up(&connection->thread_data.pending_packet);
-                    return AWS_OP_ERR;
-                }
-
-                aws_reset_error();
-                goto cleanup;
-            } else {
-                return AWS_OP_ERR;
-            }
-        }
-
-        struct aws_byte_cursor packet_data =
-            aws_byte_cursor_advance(&message_cursor, fixed_header_size + packet_header.remaining_length);
-        AWS_LOGF_TRACE(AWS_LS_MQTT_CLIENT, "id=%p: full mqtt packet read, dispatching.", (void *)connection);
-        s_process_mqtt_packet(connection, packet_header.packet_type, packet_data);
+    if (result == AWS_OP_SUCCESS) {
+        /* Do cleanup */
+        aws_channel_slot_increment_read_window(slot, message->message_data.len);
+        aws_mem_release(message->allocator, message);
```

Comment:
trivial: reverse these 2 commands. better to put the message back in the pool before you say "ok you can send me more data now".

I'm pretty sure increment_read_window() won't recurse and give you more data (but it used to).